server/src/main/java/org/apache/iotdb/db/query/dataset/DeviceIterateDataSet.java

```diff
@@ -0,0 +1,239 @@
+    deduplicatedSensorColumns = new ArrayList<>();
+    HashSet<String> tmpColumnSet = new HashSet<>();
+    for (String column : sensorColumns) {
+      if (!tmpColumnSet.contains(column)) {
+        deduplicatedSensorColumns.add(column);
+        tmpColumnSet.add(column);
+      }
+    }
+
+    if (queryPlan instanceof GroupByPlan) {
+      dataSetType = DataSetType.GROUPBY;
+      // assign parameters
+      this.expression = queryPlan.getExpression();
+      this.aggregationsGroupByDevice = ((GroupByPlan) queryPlan).getAggregationsGroupByDevice();
+      this.unit = ((GroupByPlan) queryPlan).getUnit();
+      this.origin = ((GroupByPlan) queryPlan).getOrigin();
+      this.intervals = ((GroupByPlan) queryPlan).getIntervals();
+
+    } else if (queryPlan instanceof AggregationPlan) {
+      dataSetType = DataSetType.AGGREGATE;
+      // assign parameters
+      this.aggregationsGroupByDevice = ((AggregationPlan) queryPlan).getAggregationsGroupByDevice();
+      this.expression = queryPlan.getExpression();
+
+    } else if (queryPlan instanceof FillQueryPlan) {
+      dataSetType = DataSetType.FILL;
+      // assign parameters
+      this.queryTime = ((FillQueryPlan) queryPlan).getQueryTime();
+      this.fillType = ((FillQueryPlan) queryPlan).getFillType();
+    } else {
+      dataSetType = DataSetType.QUERY;
+      // assign parameters
+      this.expression = queryPlan.getExpression();
+    }
+
+    this.curDataSetInitialized = false;
+    this.deviceIterator = pathsGroupByDevice.keySet().iterator();
+    this.curColumnsMap = new int[deduplicatedSensorColumns.size()];
+  }
+
+  public boolean hasNext() throws IOException {
+    if (curDataSetInitialized && currentDataSet.hasNext()) {
+      return true;
+    } else {
+      curDataSetInitialized = false;
+    }
+    for (int i = 0; i < deduplicatedSensorColumns.size(); i++) {
+      curColumnsMap[i] = -1;
+    }
+
+    while (deviceIterator.hasNext()) {
+      currentDevice = deviceIterator.next();
+      List<Path> paths = pathsGroupByDevice.get(currentDevice);
+      List<String> aggregations = null;
+      if (aggregationsGroupByDevice != null) {
+        aggregations = aggregationsGroupByDevice.get(currentDevice);
+      }
+
+      // get actual executed paths based on sensorColumns and get map relation
+      List<String> executeColumns = new ArrayList<>();
+      Set<String> columnSet = new HashSet<>();
+      for (int i = 0; i < paths.size(); i++) {
+        String sensor = paths.get(i).getMeasurement();
+        if (aggregations != null) {
+          columnSet.add(aggregations.get(i) + "(" + sensor + ")");
+        } else {
+          columnSet.add(sensor);
+        }
+      }
+      int index = -1;
+      for (String path : columnSet) {
+        boolean isAddToExecutePaths = false;
+        for (int i = 0; i < deduplicatedSensorColumns.size(); i++) {
+          String column = deduplicatedSensorColumns.get(i);
+          if (path.equals(column)) {
```

Comment:
I am a little confused. If `deduplicatedSensorColumns` are duplicated, how could a path equal to multiple columns in it?